lxd/auth/drivers/openfga.go

```diff
@@ -113,6 +116,83 @@ func (e *embeddedOpenFGA) load(ctx context.Context, identityCache *identity.Cach
 	return nil
 }
 
+// GetViewableProjects accepts a list of permissions and returns a list of projects that a member of a group with these permissions is able to view.
+func (e *embeddedOpenFGA) GetViewableProjects(ctx context.Context, permissions []api.Permission) ([]string, error) {
+	// Use a different context from the given context.
+	// Why? Because the OpenFGA datastore will attempt to load a cache that is optimised for permission checking an individual request.
+	// In this case we are asking the datastore to enumerate projects based on contextual tuples. The cache is not useful for this, so we don't
+	// want to trigger loading it.
+	var cancel context.CancelFunc
+	deadline, ok := ctx.Deadline()
+	if ok {
+		ctx, cancel = context.WithTimeout(context.Background(), time.Until(deadline))
+	} else {
+		ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
+	}
+
+	defer cancel()
+
+	// Construct a list objects request using a dummy identity as a member of a dummy group.
+	// Why? We're checking that, if this group *were* to exist, which projects would members of it be able to view?
+	// Neither the group nor the identity have to exist, all tuples are passed contextually.
+	// Note that we do still need to conform to our datastore implementations' expectation about the format of tuples,
+	// which is why we're passing URLs and not random strings.
+	userObject := string(entity.TypeIdentity) + ":" + entity.IdentityURL("foo", "bar").String()
+	groupObject := string(entity.TypeAuthGroup) + ":" + entity.AuthGroupURL("dummy").String()
+	req := &openfgav1.ListObjectsRequest{
+		StoreId:  dummyDatastoreULID,
+		Type:     entity.TypeProject.String(),
+		Relation: string(auth.EntitlementCanView),
+		User:     userObject,
+		ContextualTuples: &openfgav1.ContextualTupleKeys{
+			TupleKeys: []*openfgav1.TupleKey{
+				{
+					// The dummy identity is a member of the dummy group.
+					User:     userObject,
+					Relation: "member",
+					Object:   groupObject,
+				},
+			},
+		},
+	}
+
+	// Add a contextual tuple for each given permission
+	for _, permission := range permissions {
+		req.ContextualTuples.TupleKeys = append(req.ContextualTuples.TupleKeys, &openfgav1.TupleKey{
+			User:     groupObject + "#member", // Members of the dummy group have permission, not the group itself.
```

Comment:
is the "#member" here the same as the "member" string above, can these be a local constant (with "#" prepended) to show the relationship?